### PR TITLE
Check github organization membership

### PR DESCRIPTION
**Pull request by davidpaniz**
Possible solution for #10 

I extracted the method `authorize(PluginSettings, User)` to the `Provider` interface, moved the regex username validation to `GoogleProvider` and implemented the the `GitHubProvider` checking if the user is a member of the the organization.

One last thing, I was receiving a 500 error when the plugin returns 401, so I'm using a trick to not call `authenticateUser` and redirect to home.

### CODE REVIEW

File: src/main/java/com/tw/go/plugin/provider/github/GitHubProvider.java [hunk 1]
```diff
@@ -82,4 +77,42 @@ public User getUser(Profile profile) {
...
+            return isAMemberOfOrganization(pluginSettings, user);
+        }
+    }
+
+    private boolean isAMemberOfOrganization(PluginSettings pluginSettings, User user) {
+        boolean result = false;
+        try {
+            GitHub github = getGitHub(pluginSettings);
+
+            GHOrganization organization = github.getOrganization(pluginSettings.getUsernameRegex());
```
Comment (srinivasupadhya) [inline src/main/java/com/tw/go/plugin/provider/github/GitHubProvider.java:95]: this restricts `UsernameRegex` to have only one organization?
can we support multiple organizations?

File: pom.xml [hunk 2]
```diff
@@ -40,6 +40,11 @@
             <artifactId>commons-io</artifactId>
             <version>2.4</version>
         </dependency>
+        <dependency>
+            <groupId>commons-logging</groupId>
+            <artifactId>commons-logging</artifactId>
+            <version>1.1.1</version>
```
Comment (srinivasupadhya) [inline pom.xml]: you can use `com.thoughtworks.go.plugin.api.logging.Logger` & remove this dependency.

File: src/main/java/com/tw/go/plugin/provider/github/GitHubProvider.java (hunk 1 above)
Comment (davidpaniz) [inline src/main/java/com/tw/go/plugin/provider/github/GitHubProvider.java:95]: We can split by comma or use another common trick to do this, but the GH token / user password must have access to private membership of organization to call `ghUser.isMemberOf(organization)`. Another option is fallback to public information, but we should not rely on this information.
One last thing we have to keep in mind is that every verification is a http request to GH api.

I think support 1 organization is enough for most of usages and we can improve in the future if somebody ask.

Btw, we should rename the field "username_regex" if you are interest in merge this PR.

File: pom.xml (hunk 2 above)
Comment (davidpaniz) [inline pom.xml]: Actually I didn't add any logger call, but I got a kind of jar hell on gocd-server 15.2. I just tried to reproduce, but I couldn't, so I'm removing this.

File: src/main/java/com/tw/go/plugin/provider/github/GitHubProvider.java (hunk 1 above)
Comment (srinivasupadhya) [inline src/main/java/com/tw/go/plugin/provider/github/GitHubProvider.java:95]: > I think support 1 organization is enough for most of usages and we can improve in the future if somebody ask.

Ok.

> Btw, we should rename the field "username_regex" if you are interest in merge this PR.

yes im planning to have separate plugin settings for github & google, rename username_regex to organizations.

**Comment by srinivasupadhya** [comment]
thanks for this :+1:

**Pull request merged**